### PR TITLE
Drush mpx-import command

### LINKED ISSUE #7: Support storing Media subobjects in the key / value store
mpx Media objects are very complex, and don’t just contain scalars. For example, the `content` field:

- Is an array of `MediaFile` objects
  - Which itself contains an array of `releases`
    - And each `release` contains a public ID or `pid` which we want to store and load directly.

## Implementation plan
* Save MediaFiles as serialized classes in a key-value store.
* ~~At the least, store them in a separate table instead of filling up the default key / value table.~~ I've decided against this for now. The default table stores "collection" as well as an ID, and I think we can split this out later fairly easily if needed.
* Ensure related objects are deleted on entity delete.
* Consider using [entity_keyvalue](https://www.drupal.org/project/entity_keyvalue)
* Also store the Media objects themselves as a value and provide a `getMpxMedia()` method as an escape hatch for calling code in case they want the fully-typed Media object instead of going through mapped data.
* Consider wrapping the k/v store in a cache backend, if it seems like performance warrants it.

API users would call `getMetadata($media, 'content');` for example to return the array of MediaFiles, which would be retrieved from the key / value store.

## Alternatives we decided against
- Serialize into a text column in the entity table like the D7 module - too complex and fragile
- Entities for Everything (a MediaFile entity, a Release entity, etc) - Tons of work to store data we likely don't need.
- [json field](https://www.drupal.org/project/jsonb) or a JSON column in SQL - little need seen so far for querying on subobject properties, and there are database portability issues with json fields. Also, you have to manually add indexes for columns.

### LINKED ISSUE #10: Add a batch ingest drush command
We need a drush command to import existing mpx media.

* The command should take a Media type ID to determine what to import for.
* After completion, the command should tell the user to start importing through notifications instead.
* We need a way to determine if a previously imported item has been deleted.

### PR DESCRIPTION
**Pull request by deviantintegral**
Depends on #8  and resolves #10 and resolves #7 .

This PR adds initial support for importing mpx media items. I ended up removing `entity_keyvalue`. It only works once an entity has been saved (so you have an entity id) and that was leading to some super hacky workarounds. Instead, we now store media items using the id provided by mpx, instead of an entity id.

Performance is subpar with this this PR. There's no http async or batch requests - I would like to address those in followups.

To test, create an mpx Media type, and run:

```
$ vendor/bin/drush mpxi mpx_video # mpx_video is the media type ID to import for.

Importing mpx_video media
=========================

    0 [░░░░░░░░░░░░░░░░░░░░░░░░░░░░]
```

You should then see media entities in the media listing.

### CODE REVIEW

File: src/Commands/MediaMpxCommands.php [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+use Drupal\media_mpx\DataObjectFactoryCreator;
+use Drupal\media_mpx\Plugin\media\Source\Media;
+use Drush\Commands\DrushCommands;
+use Lullabot\Mpx\DataService\Access\Account;
+use Lullabot\Mpx\DataService\ByFields;
+use Lullabot\Mpx\DataService\Media\Media as MpxMedia;
+use Lullabot\Mpx\DataService\ObjectListIterator;
+
+/**
+ * A Drush commandfile.
```
Comment (juampynr) [inline src/Commands/MediaMpxCommands.php:17]: I guess that this was auto-generated.

File: src/Commands/MediaMpxCommands.php [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    $this->keyValueFactory = $keyValueFactory;
+  }
+
+  /**
+   * Import mpx objects for a media type.
+   *
+   * @param string $media_type_id
+   *   The media type ID to import for.
+   *
+   * @usage media_mpx-import mpx_video
```
Comment (juampynr) [inline src/Commands/MediaMpxCommands.php:72]: Whoa! I did not know about `@usage`.

File: src/Commands/MediaMpxCommands.php [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+      // Save the mpx Media item so it's available in getMetadata() in the
+      // source plugin.
+      $this->keyValueFactory->get('media_mpx_media')->set($mpx_media->getId(), $mpx_media);
+
+      foreach ($results as $media) {
+        $media->save();
+        $reset_ids[] = $media->id();
+      }
+
+      $this->entityTypeManager->getStorage('media')->resetCache($reset_ids);
```
Comment (juampynr) [inline src/Commands/MediaMpxCommands.php:107]: Why do you need to clear cache at this time? Are you updating results in the same loop?

File: src/Commands/MediaMpxCommands.php [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+   *   The media type object to load the source plugin for.
+   *
+   * @return \Drupal\media_mpx\Plugin\media\Source\Media
+   *   The source plugin.
+   */
+  private function loadMediaSource(MediaTypeInterface $media_type): Media {
+    /** @var \Drupal\media_mpx\Plugin\media\Source\Media $media_source */
+    $media_source = $media_type->getSource();
+    if (!($media_source instanceof Media)) {
+      throw new \RuntimeException(dt('@type is not configured as a mpx Media source.', ['@type' => $media_type->id()]));
```
Comment (juampynr) [inline src/Commands/MediaMpxCommands.php:152]: should this be `as an mpx` ?

**Comment by juampynr** [comment]
Static review looks fine and CircleCI is happy. Testing manually.

**Comment by juampynr** [comment]
Getting this:

image

Do I need to be under a VPN? Here are my account settings:

**Comment by deviantintegral** [comment]
Ah yeah, I've seen that occasionally too. Usually just rerunning drush will get the results in time.

However, I'd worked around that when working on the notification listener by disabling Drupal's default timeout entirely. I've pulled that fix back into this branch.

**Comment by juampynr** [comment]
Works now! However, see the status messages:

image

Can you fix that @deviantintegral? Then go ahead and merge this pull request.

**Pull request merged**